pkg/verifiedcar/verifiedcar.go

```diff
@@ -136,68 +135,108 @@ func (cfg Config) VerifyBlockStream(ctx context.Context, cbr BlockReader, lsys l
 		return 0, 0, traversalError(err)
 	}
 
+	if nbro.Error != nil {
+		// capture any errors not bubbled up through the traversal, i.e. see
+		// https://github.com/ipld/go-ipld-prime/pull/524
+		return 0, 0, nbro.Error
+	}
+
 	// make sure we don't have any extraneous data beyond what the traversal needs
 	_, err = cbr.Next()
-	if !errors.Is(err, io.EOF) {
+	if ctx.Err() == nil && !errors.Is(err, context.Canceled) && !errors.Is(err, io.EOF) {
 		return 0, 0, ErrExtraneousBlock
 	}
 
 	// wait for parser to finish and provide errors or stats
 	return bt.blocks, bt.bytes, nil
 }
 
-func (cfg *Config) nextBlockReadOpener(ctx context.Context, cr *carReader, bt *writeTracker, lsys linking.LinkSystem) linking.BlockReadOpener {
-	seen := make(map[cid.Cid]struct{})
-	return func(lc linking.LinkContext, l datamodel.Link) (io.Reader, error) {
-		cid := l.(cidlink.Link).Cid
-
-		var data []byte
-		var err error
-		if _, ok := seen[cid]; ok {
-			if cfg.ExpectDuplicatesIn {
-				// duplicate block, but in this case we are expecting the stream to have it
-				data, err = cr.readNextBlock(ctx, cid)
-				if err != nil {
-					return nil, err
-				}
-				if !cfg.WriteDuplicatesOut {
-					return bytes.NewReader(data), nil
-				}
-			} else {
-				// duplicate block, rely on the supplied LinkSystem to have stored this
-				rdr, err := lsys.StorageReadOpener(lc, l)
-				if !cfg.WriteDuplicatesOut {
-					return rdr, err
-				}
-				data, err = io.ReadAll(rdr)
-				if err != nil {
-					return nil, err
-				}
+type NextBlockReadOpener struct {
+	cfg  Config
+	ctx  context.Context
+	cr   *carReader
+	bt   *writeTracker
+	lsys linking.LinkSystem
+	seen map[cid.Cid]struct{}
+
+	Error error
+	Count int
+}
+
+func NewNextBlockReadOpener(
+	ctx context.Context,
+	cfg Config,
+	cr *carReader,
+	bt *writeTracker,
+	lsys linking.LinkSystem,
+) *NextBlockReadOpener {
+	return &NextBlockReadOpener{
+		cfg:  cfg,
+		ctx:  ctx,
+		cr:   cr,
+		bt:   bt,
+		lsys: lsys,
+		seen: make(map[cid.Cid]struct{}),
+	}
+}
+
+func (nbro *NextBlockReadOpener) StorageReadOpener(lc linking.LinkContext, l datamodel.Link) (io.Reader, error) {
```

Comment:
suggestion: make an internal function that just loads the block, wrap inside this function that just adds to count & sets the error if present. Will remove some repeated sets of the error and I think simplify